[apr] bump to 1.7.5 and did some imporvements, availible from this version of the library.

rinrab opened the pull request:
Bump the library to version 1.7.5 and did some imporvements, availible from this version of the library.

* Making config to be official, since it was released in the library itself.
* Fix library linkage selection.
* Disable tests to improve the build performance.
* Install include files into the directory with 'apr-1' prefix.

<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

<!-- If this PR updates an existing port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is added to each modified port's versions file.

END OF PORT UPDATE CHECKLIST (delete this line) -->

<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

rinrab committed:
[apr] bump port version and the hash of the source archive.

rinrab committed:
[apr] fix cmake config to be official

In apr of 1.7.5 version, the config was introduced. In this commit it is going to be used instead of unofficial one.

* fix-configcmake.patch: Removed.
* portfile.cmake
  (patches): Fo not apply removed fix-configcmake.patch.
  (vcpkg_cmake_config_fixup): Change package name to apr, since it is official and adjust its path.

rinrab committed:
[apr] fix library linkage selection

APR requires APR_BUILD_STATIC and APR_BUILD_SHARED options to be set for configuring the library linkage. After this commit they will be specified correctly and the library will work correctly in static configuration.

rinrab committed:
[apr] disable tests to improve the build performance.

rinrab committed:
[apr] install include files into the directory with 'apr-1' prefix

In this version it is 'recommend', but the option defaults to the previous to keep the backward compatibly.

rinrab committed:
[apr] generate new version info

Done via command: `.\vcpkg.exe x-add-version apr'.

dg0yt commented on ports/apr/portfile.cmake:
```diff
@@ -20,17 +19,24 @@ if (VCPKG_TARGET_IS_WINDOWS)
 
     vcpkg_cmake_install()
-    vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-apr CONFIG_PATH share/unofficial-apr)
+    vcpkg_cmake_config_fixup(PACKAGE_NAME "apr" CONFIG_PATH "lib/cmake/apr")
```
Nit-pick:
```suggestion
    vcpkg_cmake_config_fixup(CONFIG_PATH "lib/cmake/apr")
```
Would it make sense to polyfill unofficial-apr-config.cmake, forwarding to apr with a warning, to mitigate migration woes?
(Already did that in some ports.

rinrab replied:
I committed the suggestion (I didn't know already that it also works).

Also added unofficial-apr-config.cmake as suggested.

dg0yt replied:
The default PACKAGE_NAME is the port name.

dg0yt commented on ports/apr/portfile.cmake:
```diff
@@ -20,17 +19,24 @@ if (VCPKG_TARGET_IS_WINDOWS)
             -DINSTALL_PDB=OFF
             -DMIN_WINDOWS_VER=Windows7
             -DAPR_HAVE_IPV6=ON
+            -DAPR_INSTALL_INCLUDE_DIR=include/apr-1
```
Why is this necessary?
(No action required now.)

rinrab replied:
Okay, reverted

rinrab committed:
[apr] revert changed include directory.

rinrab committed:
[apr] do not specify the package name in config-fixup; rely it to vcpkg.

rinrab committed:
[apr] update git-tree in version info.

rinrab committed:
[apr] add config for unofficial-apr package, that forwards to APR with a warning.

* portfile.cmake
  (install): Copy unofficial-apr-config.cmake
* unofficial-apr-config.cmake: New config file for unofficial-apr package, that forwards to apr package and writes a warning.

rinrab committed:
[apr] update git-tree in version info.

rinrab committed:
[apr] follow up to commit 99cdfe639dfaad4589932aa2079a176027d51c95: fix condition by adding STREQUAL.

rinrab committed:
[apr] follow up to commit 99cdfe639dfaad4589932aa2079a176027d51c95: use INTERFACE and fake targets that links with original to redirect unofficial::apr to apr packages instead of aliases.

rinrab committed:
[apr] update git-tree in version info.

dg0yt commented on ports/apr/unofficial-apr-config.cmake:
```diff
@@ -0,0 +1,17 @@
+include(CMakeFindDependencyMacro)
+find_dependency(apr CONFIG)
+
+if("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static")
+    add_library(unofficial::apr::apr-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::apr-1 INTERFACE apr::apr-1)
+
+    add_library(unofficial::apr::aprapp-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::aprapp-1 INTERFACE apr::aprapp-1)
+else()
+    add_library(unofficial::apr::libapr-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::libapr-1 INTERFACE apr::libapr-1)
+
+    add_library(unofficial::apr::libaprapp-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::libaprapp-1 INTERFACE apr::libaprapp-1)
+endif()
```
Build-time triplet variable `VCPKG_LIBRARY_LINKAGE` is unknown at config use time, in project mode.
In addition, it is necessary to check if the target already exists.

```suggestion
if(TARGET apr::apr-1 AND NOT TARGET unofficial::apr::apr-1)
    add_library(unofficial::apr::apr-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::apr-1 INTERFACE apr::apr-1)
elseif(TARGET apr::libapr-1 AND NOT TARGET unofficial::apr::libapr-1)
    add_library(unofficial::apr::libapr-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::libapr-1 INTERFACE apr::libapr-1)
endif()
if(TARGET apr::aprapp-1 AND NOT TARGET unofficial::apr::aprapp-1)
    add_library(unofficial::apr::aprapp-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::aprapp-1 INTERFACE apr::aprapp-1)
elseif(TARGET apr::libaprapp-1 AND NOT TARGET unofficial::apr::libaprapp-1)
    add_library(unofficial::apr::libaprapp-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::libaprapp-1 INTERFACE apr::libaprapp-1)
endif()
```

I wish the upstream target name wouldn't change with library linkage, but I guess this ship has sailed.

rinrab replied:
Ops, didn't notice. Pushed the fix as https://github.com/microsoft/vcpkg/pull/40742/commits/ce0ef6f49cf3a6741a97ec63257373f5db890ee0.

I think there should be the check for the target existing, because otherwise the target will be created, but will work incorrectly.

rinrab committed:
[apr] fix unofficial-apr target creation

Check for target existing instead of using VCPKG_LIBRARY_LINKAGE variable for creation 'alias' targets.

rinrab committed:
[apr] update git-tree in version info.

JonLiu1993 commented:
@rinrab, Thanks for your pr, when I tested the usage by `apr:x64-windows` , I get this error,please take a look:
````
1> CMake generation started for default configuration: 'x64-Debug'.
1> Found and using vcpkg toolchain file (F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake).
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"
1> Working directory: C:\Users\test\source\repos\CMakeProject1\out\build\x64-Debug
1> [CMake] -- The C compiler identification is MSVC 19.40.33814.0
1> [CMake] -- The CXX compiler identification is MSVC 19.40.33814.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] CMake Warning at F:/Feature-test/apr/installed/x64-windows/share/unofficial-apr/unofficial-apr-config.cmake:1 (message):
1> [CMake]   find_package(unofficial-apr) is deprecated.
1> [CMake] 
1> [CMake]   Use find_package(apr) instead
1> [CMake] Call Stack (most recent call first):
1> [CMake]   F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   CMakeProject1/CMakeLists.txt:6 (find_package)
1> [CMake] 
1> [CMake] 
1> [CMake] -- Configuring done (2.0s)
1> [CMake] CMake Error at CMakeProject1/CMakeLists.txt:10 (target_link_libraries):
1> [CMake]   Target "test" links to:
1> [CMake] 
1> [CMake]     unofficial::apr::apr-1
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
1> [CMake] 
1> [CMake] 
1> [CMake] 
1> [CMake] -- Generating done (0.0s)
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"' returned with exit code: 1'.

````
CMakeLists.txt:
````
cmake_minimum_required(VERSION 3.5)

project(test)

find_package(unofficial-apr CONFIG REQUIRED)

add_executable(test CMakeProject1.cpp)

target_link_libraries(test PRIVATE unofficial::apr::apr-1 unofficial::apr::aprapp-1 unofficial::apr::libapr-1 unofficial::apr::libaprapp-1)
````

rinrab commented:
This library has non-standard targets. By default, APR builds both static and shared configurations. The target names and library artifacts are `apr-1` for the static configuration and `libapr-1` for the shared configuration. However, either of these configurations can be disabled via options (which done by the port).

So, when you install the library for the `x64-windows` triplet, you only get the shared library (`libapr-1` and `aprapp-1`). This results in the non-existence of `apr::apr-1`, `apr::aprapp-1`, and their unofficial versions.

If you remove linking against `unofficial::apr::apr-1` and `unofficial::apr::aprapp-1` from your CMakeLists, it should work.

rinrab committed:
[apr] do not specify port-version, since it is first for the new version

rinrab committed:
[apr] regenerate version info

rinrab commented on ports/apr/vcpkg.json:
```diff
@@ -1,6 +1,6 @@
   "name": "apr",
-  "version": "1.7.4",
+  "version": "1.7.5",
   "port-version": 1,
```
Thanks! I forgot to remove it.

JonLiu1993 commented:
Tested usage successfully by `apr:x64-windows`:
````
apr provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(apr CONFIG REQUIRED)
  target_link_libraries(main PRIVATE apr::libapr-1 apr::libaprapp-1)

  find_package(unofficial-apr CONFIG REQUIRED)
  target_link_libraries(main PRIVATE unofficial::apr::apr-1 unofficial::apr::aprapp-1 unofficial::apr::libapr-1 unofficial::apr::libaprapp-1)
````

dg0yt commented:
> ~~~
> target_link_libraries(main PRIVATE unofficial::apr::apr-1 unofficial::apr::aprapp-1 unofficial::apr::libapr-1 unofficial::apr::libaprapp-1)

Ouch. Why does this form of testing and feedback never stop?

Like all port with varying exported target names, this port should get a `usage` file with the `target_link_libraries(... $<IF:$<TARGET_EXISTS:...>,...,...>)` pattern.

rinrab commented:
Opened another pull-request with adding the correct usage.
